packages/datasets-fs/src/commands/generate.ts

```diff
@@ -0,0 +1,169 @@
+        if (unlabelledMapFile && !fs.existsSync(unlabelledMapFile)) {
+            throw new ProsopoEnvError(
+                new Error(`unlabelled map file does not exist: ${unlabelledMapFile}`),
+                'FS.FILE_NOT_FOUND'
+            )
+        }
+        this.labelledMapFile = labelledMapFile || ''
+        this.unlabelledMapFile = unlabelledMapFile || ''
+    }
+
+    labelled: LabelledItem[] = []
+    unlabelled: Item[] = []
+    labels: string[] = []
+    labelledMapFile = ''
+    unlabelledMapFile = ''
+    labelToImages: { [label: string]: Item[] } = {}
+    targets: string[] = []
+    saltRounds = 10
+
+    public override async _run(args: Args) {
+        await super._run(args)
+        // set the seed
+        setSeedGlobal(args.seed || 0)
+        // get lodash (with seeded rng)
+        const _ = lodash()
+
+        const allowDuplicatesLabelled = args.allowDuplicatesLabelled || args.allowDuplicates || false
+        const allowDuplicatesUnlabelled = args.allowDuplicatesUnlabelled || args.allowDuplicates || false
+
+        // load the map to get the labelled and unlabelled data
+        this.labelled = this.labelledMapFile
+            ? LabelledDataSchema.parse(JSON.parse(fs.readFileSync(this.labelledMapFile, 'utf8'))).items
+            : []
+        this.unlabelled = this.unlabelledMapFile
+            ? DataSchema.parse(JSON.parse(fs.readFileSync(this.unlabelledMapFile, 'utf8'))).items
+            : []
+
+        // check for duplicates
+        checkDuplicates(this.labelled, this.unlabelled, {
+            allowDuplicatesLabelled,
+            allowDuplicatesUnlabelled,
+        })
+
+        // split the labelled data by label
+        this.labelToImages = {}
+        for (const entry of this.labelled) {
+            const arr = this.labelToImages[entry.label] || []
+            arr.push(entry)
+            this.labelToImages[entry.label] = arr
+        }
+        this.targets = Object.keys(this.labelToImages)
+
+        // load the labels from file
+        // these are the labels that unlabelled data will be assigned to
+        // note that these can be different to the labels in the map file as the labelled data is independent of the unlabelled data in terms of labels
+        this.labels = []
+        if (args.labels && fs.existsSync(args.labels)) {
+            this.labels.push(
+                ...[...LabelsContainerSchema.parse(JSON.parse(fs.readFileSync(args.labels, 'utf8'))).labels]
+            )
+        } else {
+            // else default to the labels in the labelled data
+            this.labels.push(...[...this.targets])
+        }
+    }
+}
+
+export const checkDuplicates = (
+    labelled: LabelledItem[],
+    unlabelled: Item[],
+    options: {
+        allowDuplicatesLabelled?: boolean
+        allowDuplicatesUnlabelled?: boolean
+    }
+) => {
+    // check for duplicates
+    const all = new Set<string>()
+    if (!options.allowDuplicatesLabelled) {
```

Comment:
Error handling here could be shared function, with only all.add changing depending on labelled/unlabelled